src/plugins/clips-executive/clips/domain.clp

```diff
@@ -216,263 +633,188 @@
   (modify ?a (param-names ?param-names))
 )
 
-(defrule domain-add-operator-to-top-precondition
-  "If a precondition has no operator but is part of a name that is an operator,
-   add that name as operator to the precondition."
-  ?precond <- (domain-precondition (operator nil) (part-of ?op))
-  (domain-operator (name ?op))
-=>
-  (modify ?precond (operator ?op))
-)
-
-(defrule domain-add-operator-to-nested-precondition
-  "If a precondition does not have an operator but the parent has one, then copy
-   the operator from the parent to the child."
-  ?precond <- (domain-precondition (operator nil) (part-of ?parent))
-  (domain-precondition (name ?parent) (operator ?op&~nil))
-=>
-  (modify ?precond (operator ?op))
-)
-
-(defrule domain-add-operator-to-atomic-precondition
-  "If an atomic precondition does not have an operator but the parent has one,
-   then copy the operator from the parent to the child."
-  ?precond <- (domain-atomic-precondition (operator nil) (part-of ?parent))
-  (domain-precondition (name ?parent) (operator ?op&~nil))
-=>
-  (modify ?precond (operator ?op))
-)
-
 (deffunction remove-precondition
-  "Remove an atomic precondition from its parent and clean up the precondition
-   tree. If the parent is a disjunction with no other disjunct, simplify it to
+  "Remove a sub-formula from its parent and clean up the formula tree.
+   If the parent is a disjunction with no other disjunct, simplify it to
    true by removing it recursively. If it is a negation, remove it recursively.
    If it's a conjunction, only remove the conjunct.
-   If the top-most precondition is removed, replace it by a trivially true one
+   If the top-most formula node is removed, replace it by a trivially true one
    (empty conjunction)."
   (?precond-fact)
-  (if (not (do-for-fact ((?parent domain-precondition))
-               (eq (fact-slot-value ?precond-fact part-of) ?parent:name)
-    (if (or (eq (fact-slot-value ?parent type) negation)
-            (and (eq (fact-slot-value ?parent type) disjunction)
-                 (not (any-factp ((?sibling domain-precondition))
-                       (eq (fact-slot-value ?parent name) ?sibling:part-of)))
-                 (not (any-factp ((?sibling domain-atomic-precondition))
-                       (eq (fact-slot-value ?parent name) ?sibling:part-of)))))
-     then
+
+  (bind ?parent FALSE)
+  (do-for-fact ((?p-fact pddl-formula))
+               (eq (fact-slot-value ?precond-fact part-of) ?p-fact:id)
+    (bind ?parent ?p-fact)
+  )
+
+  (if ?parent then
```

Comment:
Why not just
```suggestion
  (do-for-fact ((?parent pddl-formula))
               (eq (fact-slot-value ?precond-fact part-of) ?parent:id)
  )
```

Reply:
I think I couldn't make the distinction between formulas that have a parent and those that don't anymore.

Reply:
`do-for-fact` returns `FALSE` if no fact exists:
> If no fact-set satisfied the query, then the return value
is the symbol FALSE.

Does that help?